test/profiling/alt-impl/cid2.js

```diff
@@ -0,0 +1,275 @@
+      multihash.prefix(this.multihash)
+    ])
+  }
+
+  /**
+   * Convert to a CID of version `0`.
+   *
+   * @returns {CID2}
+   */
+  toV0 () {
+    if (this.codec !== 'dag-pb') {
+      throw new Error('Cannot convert a non dag-pb CID to CIDv0')
+    }
+
+    return new CID2(0, this.codec, this.multihash)
+  }
+
+  /**
+   * Convert to a CID of version `1`.
+   *
+   * @returns {CID2}
+   */
+  toV1 () {
+    return new CID2(1, this.codec, this.multihash)
+  }
+
+  /**
+   * Encode the CID into a string.
+   *
+   * @param {string} [base='base58btc'] - Base encoding to use.
+   * @returns {string}
+   */
+  toBaseEncodedString (base) {
+    base = base || 'base58btc'
+
+    switch (this.version) {
+      case 0: {
+        if (base !== 'base58btc') {
+          throw new Error('not supported with CIDv0, to support different bases, please migrate the instance do CIDv1, you can do that through cid.toV1()')
+        }
+        return mh.toB58String(this.multihash)
+      }
+      case 1:
+        return multibase.encode(base, this.buffer).toString()
+      default:
+        throw new Error('Unsupported version')
+    }
+  }
+
+  /**
+   * Serialize to a plain object.
+   *
+   * @returns {SerializedCID}
+   */
+  toJSON () {
+    return {
+      codec: this.codec,
+      version: this.version,
+      hash: this.multihash
+    }
+  }
+
+  /**
+   * Compare equality with another CID.
+   *
+   * @param {CID2} other
+   * @returns {bool}
+   */
+  equals (other) {
+    return this.codec === other.codec &&
+      this.version === other.version &&
+      this.multihash.equals(other.multihash)
+  }
+
+  /**
+   * Test if the given input is a CID.
+   *
+   * @param {any} other
+   * @returns {bool}
+   */
+  static isCID (other) {
+    try {
+      let errorMsg = CID2._checkCIDComponents(other)
+      return !(errorMsg)
+    } catch (err) {
+      return false
+    }
+  }
+
+  /**
+   * Test if the given input is a valid CID object.
+   * Throws if it is not.
+   *
+   * @param {any} other
+   * @returns {void}
+   */
+  static validateCID (other) {
+    let errorMsg = CID2._checkCIDComponents(other)
+    if (errorMsg) {
+      throw new Error(errorMsg)
+    }
+  }
+
+  /**
+   * Test if the given input is a valid CID object.
+   * Returns an error message if it is not, or
+   * Throws an error its multihash is invalid.
+   *
+   * @param {any} other
+   * @returns {string}
+   */
+  static _checkCIDComponents (other) {
+    if (other == null) {
+      return 'null values are not valid CIDs'
+    }
+
+    if (!(other.version === 0 || other.version === 1)) {
+      return 'Invalid version, must be a number equal to 1 or 0'
+    }
+
+    if (typeof other.codec !== 'string') {
+      return 'codec must be string'
+    }
+
+    if (!Buffer.isBuffer(other.multihash)) {
+      return 'multihash must be a Buffer'
+    }
+
+    mh.validate(other.multihash)
+    return null
+  }
+}
+
+CID2.codecs = codecs
+
+module.exports = CID2
```

Comment:
Why duplicate instead of using the version in src?

Reply:
@diasdavid Thanks for pointing that out. I wasn't really happy with this solution, but I couldn't come up with a better one. The issue was: (1) I wanted to use, _as a reference,_ a "snapshot" of the CID implementation as of, say, Sunday. The CID implementation will (likely) change over time, and using the "current" version (in src) will not make sense for this particular test (say, 2 months from now). (2) I wanted to try a different "alternative" implementations _for comparison._ And, putting alternative (trial) implementations in the prod source code (e.g., as an extra method, etc.) would not have made sense. 

That's why I ended up duplicating the code for _both._ Obviously, there is something "funny" about doing it this way. I'm not sure what would be the best way though. Any suggestions?

Reply:
In hindsight, I think your suggestion, if I understand correctly, seems to make more sense. I now have to reverse reference and alternative (in my mind):

* Reference: CID version X (in NPM)
* Alternative: CID in src (with alternative/trial implementation)

(When I started out, the CID in src was the reference, and "my version" was the alternative.)

Reply:
You can npm install a locked version on the profiling test folder (therefore one version) and require the other from the src folder (and you get the other version to compare). Makes sense?